sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -154,6 +274,21 @@ private[sql] case class PreWriteCheck(conf: SQLConf, catalog: SessionCatalog)
 
   def apply(plan: LogicalPlan): Unit = {
     plan.foreach {
+      case c @ CreateTable(tableDesc, mode, query) if c.resolved =>
+        // Since we are saving table metadata to metastore, we should make sure the table name
+        // and database name don't break some common restrictions, e.g. special chars except
+        // underscore are not allowed.
+        val pattern = Pattern.compile("[\\w_]+")
```

Comment:
Compiling a pattern can be expensive and the `Pattern` produced is reusable; that is all.